runtime/src/main/cpp/Memory.cpp

```diff
@@ -635,20 +753,72 @@ inline void DecrementRC(ContainerHeader* container) {
       container->setColorAssertIfGreen(CONTAINER_TAG_GC_PURPLE);
       if (!container->buffered()) {
         container->setBuffered();
-        auto state = memoryState;
         state->toFree->push_back(container);
-        if (state->gcSuspendCount == 0 && freeableSize(state) >= state->gcThreshold) {
-          GarbageCollect();
+        GC_LOG("toFree is now %d\n", state->toFree->size());
+        if (state->gcSuspendCount == 0 && state->toRelease->size() >= state->gcThreshold) {
+          GC_LOG("Calling GC from DecrementRC: %d\n", state->toRelease->size())
+          garbageCollect(state, false);
         }
       }
     }
   }
 }
 
-inline void initThreshold(MemoryState* state, uint32_t gcThreshold) {
+inline void DecrementRC(ContainerHeader* container) {
+  auto* state = memoryState;
+  // TODO: enable me, once account for inner references in frozen objects correctly.
+  // RuntimeAssert(container->refCount() > 0, "Must be positive");
+  bool useCycleCollector = container->tag() == CONTAINER_TAG_NORMAL;
+  if (container->decRefCount() == 0) {
+    freeContainer(container);
+  } else if (useCycleCollector && state->toFree != nullptr) {
+      RuntimeAssert(container->refCount() > 0, "Must be positive");
+      RuntimeAssert(!container->shareable(), "Cycle collector shalln't be used with shared objects yet");
+      RuntimeAssert(container->objectCount() == 1, "cycle collector shall only work with single object containers");
+      // We do not use cycle collector for frozen objects, as we already detected
+      // possible cycles during freezing.
+      // Also do not use cycle collector for provable acyclic objects.
+      int color = container->color();
+      if (color != CONTAINER_TAG_GC_PURPLE && color != CONTAINER_TAG_GC_GREEN) {
+        container->setColorAssertIfGreen(CONTAINER_TAG_GC_PURPLE);
+        if (!container->buffered()) {
+          container->setBuffered();
+          state->toFree->push_back(container);
+          if (state->gcSuspendCount == 0 && state->toFree->size() >= state->gcThreshold) {
+              GC_LOG("Calling GC from DecrementRC: %d\n", state->toRelease->size())
+              garbageCollect(state, false);
```

Comment:
This `DecrementRC` variant is used only during GC. What's the purpose of invoking GC inside it?